added additional logic to contact_topic_answers to improve discoverab…

josephmsmith opened the pull request:
### What github issue is this PR for, if any?
Resolves #5775 

### What changed, and _why_?
I added additional logic to contact_topic_answers to improve discoverability for case topics if admin has not set any court report topics. This was important because users did not know admins needed to set court report topics in order for the feature to be accessible by volunteers, supervisors, and admins.

### How is this **tested**? (please write tests!) 💖💪
This was tested by removing all topics from Court Report Topics as an admin, then logging in as each user type and confirming what verbiage was displayed as it needed to be different for admins versus volunteers and supervisors. 

I then logged in as an admin then set court report topics and repeated the process of logging in as each user type to confirm court report topics would be visible.

For each test, I inspected the code and simulated being in a different browser type (mobile, tablet, laptop) to confirm the CSS classes conformed with the rest of the page.\

### Screenshots please :)
![court_report_NO_topics_admin_laptop](https://github.com/rubyforgood/casa/assets/121319535/a270a436-612f-478b-93ff-a614145548f6)
![court_report_NO_topics_mobile_admin](https://github.com/rubyforgood/casa/assets/121319535/5952c19e-974a-43cb-b2c1-5914361aee6c)
![court_report_NO_topics_volunteer_laptop](https://github.com/rubyforgood/casa/assets/121319535/5143ac7b-9350-4412-abd3-cad8c2cbb340)
![court_report_NO_topics_volunteer_mobile](https://github.com/rubyforgood/casa/assets/121319535/ebfb9284-8618-436e-810c-f7dd238ada39)
![court_report_topic_mobile_admin](https://github.com/rubyforgood/casa/assets/121319535/fc1268fd-07b2-480a-b71c-eef0f4fcd0bd)
![court_report_topic_tablet_admin](https://github.com/rubyforgood/casa/assets/121319535/1c097b47-7703-4393-ab30-db0aba9d79f6)
![court_report_topics_laptop_supervisor](https://github.com/rubyforgood/casa/assets/121319535/5311dce7-eccb-4502-bd4e-6fdcb68942fa)


### Feelings gif (optional)
_What gif best describes your feeling working on this issue? https://giphy.com/
`![noice](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdjBmMmFrd25wMmI5ZjAyZjU3dWljcXNkYTMzc3h0ZTN2bnJqZHNndyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/4KF85OSbyjVOfyjksJ/giphy.gif)`

josephmsmith committed:
added additional logic to contact_topic_answers to improve discoverability for case topics if admin has not set any court report topics

compwron reviewed (commented):
Looks promising! I see an erb lint failure. Please add a test :)

josephmsmith committed:
updating form_spec.rb with volunteer sign in tests

josephmsmith committed:
ran erb lint and added test cases to form_spec to ensure testing when different users view the page and no contact topics are available

josephmsmith committed:
run standardrb on form_spec.rb

josephmsmith commented:
@compwron can you help guide on next steps for the two failures?

elasticspoon commented:
> @compwron can you help guide on next steps for the two failures?

Those are false positives. Unfortunately there are issues with some tests timing out and looking like failures. 

So your pr passes the test suite.

elasticspoon commented on app/views/case_contacts/form/details.html.erb:
```diff
@@ -78,18 +78,28 @@
-            <%= f.check_box :selected, class: ["form-check-input", "casa-case-id"] %>
-            <%= f.label :selected, f.object.contact_topic.question, class: "form-check-label" %>
-          </div>
+      <div class="card-style-1 pl-25 pr-25 mb-10">
+        <h4 class="mb-20 details__topics-label">Court report <%= "topic".pluralize(@case_contact.contact_topic_answers.count) %> <span class="content-1">(optional)</span></h4>
+        <div>
+          <%= form.fields_for(:contact_topic_answers) do |f| %>
+            <div class="form-check checkbox-style mb-10 ml-5">
+              <%= f.check_box :selected, class: ["form-check-input", "casa-case-id"] %>
+              <%= f.label :selected, f.object.contact_topic.question, class: "form-check-label" %>
+            </div>
+          <% end %>
+        </div>
+      </div>
+    <% elsif @case_contact.contact_topic_answers.none? %>
+      <div class="card-style-1 pl-25 pr-25 mb-10">
+        <h4 class="mb-20 details__topics-label">Court Report Topics</h4>
+        <% if current_user.volunteer? || current_user.supervisor? %>
+          <div class="mb-10 ml-5">Your organization has not set any Court Report Topics yet. Contact your admin to learn more.</div>
+        <% elsif current_user.casa_admin? %>
+          <div>Visit <%= link_to "Manage Case Contact Topics", edit_casa_org_path(current_user.casa_org_id) %> to set your organization Court report topics.</div>
```
This seems like a lot of duplication of HTML. I think we can make the if statement only dictate this part:

```erb
  <%= form.fields_for(:contact_topic_answers) do |f| %>
    <div class="form-check checkbox-style mb-10 ml-5">
      <%= f.check_box :selected, class: ["form-check-input", "casa-case-id"] %>
      <%= f.label :selected, f.object.contact_topic.question, class: "form-check-label" %>
    </div>
  <% end %>
```

Basically
- if there are topics show that part
- if there are none use the admin/volunteer logic you have.
- lets make the volunteer text be the default. you can make that the else clause instead of using if / elsif

I also think I like this as the heading more than what I had before:

```erb
 <h4 class="mb-20 details__topics-label">Court Report Topics <span class="content-1">(optional)</span></h4>
```

josephmsmith replied:
Adjusted! Thank you for the suggestions - let me know if you see something that doesn't look right.

elasticspoon reviewed (changes requested):
Nice tests! This is pretty close to what we are looking for. See comments for suggestions.

josephmsmith committed:
wip

josephmsmith committed:
fix for linted code

josephmsmith committed:
final linted

josephmsmith committed:
cleaned up previous linting problems with other linter
, moved volunteer and supervisor tests under GET /show to improve future test writing

josephmsmith committed:
additional linting for casa_case

josephmsmith committed:
refactored the html to eliminate redundancy, adhere to casa styling requirements, and simplify logic

josephmsmith commented on spec/requests/case_contacts/form_spec.rb:
Done!

josephmsmith commented on spec/requests/case_contacts/form_spec.rb:
```diff
@@ -417,4 +432,41 @@
       end
     end
   end
+  # TODO
```
TODONE!

elasticspoon reviewed (approved):
LGTM! Nice job!

compwron commented:
Looks good! Sorry about the flaky spec. 
```

  1) case_contacts/index with case contacts without filter displays the contact type groups
     Failure/Error: expect(page).to have_text(case_contacts[2].contact_groups_with_types.keys.first)
       expected to find text "Group 1388" in "Group 1389\nDraft"
```

elasticspoon commented:
@josephmsmith this is super minor thing but FYI
in git commits you typically want to write in the first line main point of the commit in 80 chars or less.
then have an empty line 
the write the body of the commit which has move context.

you have a few commits where the wrote a really long first line of they look like 
![image](https://github.com/rubyforgood/casa/assets/14540596/70c589e4-42d6-469a-8cd3-5a222844b16a)
vs 

![image](https://github.com/rubyforgood/casa/assets/14540596/7f3e46d1-de22-46fa-ad75-f051a298ad82)

josephmsmith commented:
Noted! Thank you for all your help @elasticspoon
